cmd/cosign/cli/verify/verify_bundle.go

```diff
@@ -149,15 +164,159 @@ func verifyNewBundle(ctx context.Context, bundlePath, trustedRootPath, keyRef, s
+		MediaType:            "application/vnd.dev.sigstore.bundle+json;version=0.3",
+		VerificationMaterial: &protobundle.VerificationMaterial{},
+	}
+
+	if envelope != nil && len(envelope.Signatures) > 0 {
+		sigDecode, err := base64.StdEncoding.DecodeString(envelope.Signatures[0].Sig)
+		if err != nil {
+			return nil, err
+		}
+
+		sig := &protodsse.Signature{
+			Sig: sigDecode,
+		}
+
+		payloadDecode, err := base64.StdEncoding.DecodeString(envelope.Payload)
+		if err != nil {
+			return nil, err
+		}
+
+		pb.Content = &protobundle.Bundle_DsseEnvelope{
+			DsseEnvelope: &protodsse.Envelope{
+				Payload:     payloadDecode,
+				PayloadType: envelope.PayloadType,
+				Signatures:  []*protodsse.Signature{sig},
+			},
+		}
+	} else {
+		pb.Content = &protobundle.Bundle_MessageSignature{
+			MessageSignature: &protocommon.MessageSignature{
+				MessageDigest: &protocommon.HashOutput{
+					Algorithm: protocommon.HashAlgorithm_SHA2_256,
+					Digest:    digest[:],
+				},
+				Signature: sigBytes,
+			},
+		}
+	}
+
+	if cert != nil {
+		pb.VerificationMaterial.Content = &protobundle.VerificationMaterial_Certificate{
+			Certificate: &protocommon.X509Certificate{
+				RawBytes: cert.Raw,
+			},
+		}
+	} else if sigVerifier != nil {
+		pub, err := sigVerifier.PublicKey(pkOpts...)
+		if err != nil {
+			return nil, err
+		}
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(pub)
+		if err != nil {
+			return nil, err
+		}
+		hashedBytes := sha256.Sum256(pubKeyBytes)
+
+		pb.VerificationMaterial.Content = &protobundle.VerificationMaterial_PublicKey{
+			PublicKey: &protocommon.PublicKeyIdentifier{
+				Hint: base64.StdEncoding.EncodeToString(hashedBytes[:]),
+			},
+		}
+	}
+
+	if len(signedTimestamp) > 0 {
+		ts := &protocommon.RFC3161SignedTimestamp{
+			SignedTimestamp: signedTimestamp,
+		}
+
+		pb.VerificationMaterial.TimestampVerificationData = &protobundle.TimestampVerificationData{
+			Rfc3161Timestamps: []*protocommon.RFC3161SignedTimestamp{ts},
+		}
+	}
+
+	if !ignoreTlog {
+		var pem []byte
+		var err error
+		if cert != nil {
+			pem, err = cryptoutils.MarshalCertificateToPEM(cert)
+			if err != nil {
+				return nil, err
+			}
+		} else if sigVerifier != nil {
+			pub, err := sigVerifier.PublicKey(pkOpts...)
+			if err != nil {
+				return nil, err
+			}
+			pem, err = cryptoutils.MarshalPublicKeyToPEM(pub)
+			if err != nil {
+				return nil, err
+			}
+		}
+		var sigB64 string
+		var payload []byte
+		if envelope != nil && len(envelope.Signatures) > 0 {
+			payload, err = json.Marshal(*envelope)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			sigB64 = base64.StdEncoding.EncodeToString(sigBytes)
+			payload = buf.Bytes()
+		}
+
+		tlogEntries, err := cosign.FindTlogEntry(ctx, rekorClient, sigB64, payload, pem)
+		if err != nil {
+			return nil, err
+		}
+		if len(tlogEntries) == 0 {
+			return nil, fmt.Errorf("unable to find tlog entry")
+		}
+		// Attempt to verify with the earliest integrated entry
```

Comment:
Since `FindTlogEntry` is searching based on entry hash, you should only match one entry, so you could add that as a check.

Reply:
I was copying https://github.com/sigstore/cosign/blob/9b17791a21f9eef8ff59e6ea138e1b116dc571b8/pkg/cosign/verify.go#L446-L469, but this sounds fine to me - I updated the implementation to check for exactly one entry.